src/AspNetCoreModuleV2/AspNetCore/HandlerResolver.cpp

```diff
@@ -17,129 +17,140 @@ const PCWSTR HandlerResolver::s_pwzAspnetcoreOutOfProcessRequestHandlerName = L"
+            location = options->GetExeLocation();
 
-            if (FAILED_LOG(hr = FindNativeAssemblyFromHostfxr(options.get(), pstrHandlerDllName, &struFileName)))
+            if (FAILED_LOG(hr = FindNativeAssemblyFromHostfxr(options.get(), pstrHandlerDllName, handlerDllPath)))
             {
                 EventLog::Error(
                     ASPNETCORE_EVENT_INPROCESS_RH_MISSING,
                     ASPNETCORE_EVENT_INPROCESS_RH_MISSING_MSG,
-                    struFileName.IsEmpty() ? s_pwzAspnetcoreInProcessRequestHandlerName : struFileName.QueryStr());
+                    handlerDllPath.empty() ? s_pwzAspnetcoreInProcessRequestHandlerName : handlerDllPath.c_str());
 
                 return hr;
             }
         }
         else
         {
-            if (FAILED_LOG(hr = FindNativeAssemblyFromGlobalLocation(pConfiguration, pstrHandlerDllName, &struFileName)))
+            if (FAILED_LOG(hr = FindNativeAssemblyFromGlobalLocation(pConfiguration, pstrHandlerDllName, handlerDllPath)))
             {
                 EventLog::Error(
                     ASPNETCORE_EVENT_OUT_OF_PROCESS_RH_MISSING,
                     ASPNETCORE_EVENT_OUT_OF_PROCESS_RH_MISSING_MSG,
-                    struFileName.IsEmpty() ? s_pwzAspnetcoreOutOfProcessRequestHandlerName : struFileName.QueryStr());
+                    handlerDllPath.empty() ? s_pwzAspnetcoreOutOfProcessRequestHandlerName : handlerDllPath.c_str());
 
                 return hr;
             }
         }
 
-        LOG_INFOF("Loading request handler:  %S", struFileName.QueryStr());
+        LOG_INFOF("Loading request handler:  %S", handlerDllPath.c_str());
 
-        m_hRequestHandlerDll = LoadLibraryW(struFileName.QueryStr());
-        RETURN_LAST_ERROR_IF_NULL(m_hRequestHandlerDll);
+        hRequestHandlerDll = LoadLibrary(handlerDllPath.c_str());
+        if (preventUnload)
+        {
+            // Pin module in memory
+            GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_PIN, pstrHandlerDllName, &hRequestHandlerDll);
+        }
+        RETURN_LAST_ERROR_IF_NULL(hRequestHandlerDll);
     }
 
-    m_pfnAspNetCoreCreateApplication = reinterpret_cast<PFN_ASPNETCORE_CREATE_APPLICATION>(GetProcAddress(m_hRequestHandlerDll, "CreateApplication"));
-
-    RETURN_LAST_ERROR_IF_NULL(m_pfnAspNetCoreCreateApplication);
+    auto pfnAspNetCoreCreateApplication = reinterpret_cast<PFN_ASPNETCORE_CREATE_APPLICATION>(GetProcAddress(hRequestHandlerDll, "CreateApplication"));
+    RETURN_LAST_ERROR_IF_NULL(pfnAspNetCoreCreateApplication);
 
+    pApplicationFactory = std::make_unique<ApplicationFactory>(hRequestHandlerDll.release(), location, pfnAspNetCoreCreateApplication);
     return S_OK;
 }
 
 HRESULT
-HandlerResolver::GetApplicationFactory(IHttpApplication &pApplication, STRU& location, PFN_ASPNETCORE_CREATE_APPLICATION * pfnCreateApplication)
+HandlerResolver::GetApplicationFactory(IHttpApplication &pApplication, std::unique_ptr<ApplicationFactory>& pApplicationFactory)
```

Comment:
Function name is confusing, you aren't getting the factory at all.

Reply:
You are getting ApplicationFactory object

Reply:
I see now, my mistake. This is more of an initialize method though. with a call to LoadRequestHandlerAssembly. Can you move the initialize into the constructor?

Reply:
Why is it initialize? It's called per application_info creation

Reply:
I see now, the HandlerResolver object is part of the application manager, so you never call new HandlerResolver.